lib/puppet/provider/cs_stonith/pcs.rb

```diff
@@ -0,0 +1,209 @@
+  end
+
+  # Unlike create we actually immediately delete the item.
+  def destroy
+    debug('Removing STONITH fence')
+    Puppet::Provider::Pacemaker::run_pcs_command([command(:pcs), 'stonith', 'delete', @property_hash[:name]])
+    @property_hash.clear
+  end
+
+  # Getters that obtains the device_options and operations defined in our primitive
+  # that have been populated by prefetch or instances (depends on if your using
+  # puppet resource or not).
+  def primitive_type
+    @property_hash[:primitive_type]
+  end
+
+  def device_options
+    @property_hash[:device_options]
+  end
+
+  def operations
+    @property_hash[:operations]
+  end
+
+  # Our setters for device_options and operations.  Setters are used when the
+  # resource already exists so we just update the current value in the
+  # property_hash and doing this marks it to be flushed.
+  def fence_type=(should)
+    @property_hash[:fence_type] = should
+  end
+
+  def device_options=(should)
+    @property_hash[:device_options] = should
+  end
+
+  def operations=(should)
+    @property_hash[:operations] = should
+  end
+
+  # Flush is triggered on anything that has been detected as being
+  # modified in the property_hash.
+  # It calls several pcs commands to make the resource look like the
+  # params.
+  def flush
+    unless @property_hash.empty?
+      # The primitive_type variable is used to check if one of the class,
+      # provider or type has changed
+      primitive_type = "#{@property_hash[:primitive_type]}"
+
+      unless @property_hash[:device_options].empty?
+        device_options = []
+        @property_hash[:device_options].each_pair do |k,v|
+          device_options << "#{k}=#{v}"
+        end
+      end
+
+      unless @property_hash[:operations].empty?
+        operations = []
+        @property_hash[:operations].each do |o|
+          op_name = o[0]
+          operations << "op"
+          if op_name.include? ':'
+            items = op_name.split(':')
+            operations << items[0]
+            operations << "role=#{items[1]}"
+          else
+            operations << op_name
+          end
+          o[1].each_pair do |k,v|
+            operations << "#{k}=#{v}"
+          end
+        end
+      end
+
+      # We destroy the resource if it's primitive type has changed
+      unless @property_hash[:existing_resource] == :false
+        existing_primitive_type = "#{@property_hash[:existing_primitive_type]}"
+        if existing_primitive_type != primitive_type
+          debug('Removing fence device')
+          Puppet::Provider::Pacemaker::run_pcs_command([command(:pcs), 'stonith', 'delete', @property_hash[:name]])
+          force_reinstall = :true
+        end
+      end
+
+      if @property_hash[:existing_resource] == :false or force_reinstall == :true
+        cmd = [ command(:pcs), 'stonith', 'create', "#{@property_hash[:name]}" ]
+        cmd << primitive_type
+        cmd += device_options unless device_options.nil?
```

Comment:
`<<` and `+=` have different semantics; `<<` is `#push` and `+=` is `#concat`. In this case it might be more clear to use the actual methods instead of the overloaded operators.